README.md
# SecureABC Paper and Reference Implementation

## Paper
The latest SecureABC AntiBody Certificates paper can be found at [SecureABC.pdf](SecureABC.pdf). The authors welcome feedback and may be contacted by email using the details in the paper or by clicking the following [link](mailto:chicks@turing.ac.uk;dbutler@turing.ac.uk;cm@warwick.ac.uk;jon.crowcroft@cl.cam.ac.uk?subject=[SecureABC]).

Changes to the paper are documented in the [changelog](changelog.md).

## Reference Implementation

This repository also includes a simple reference implementation (in Python3), and proof-of-concept, of the SecureABC AntiBody Certificate proposal. In particular we currently demonstrate the feasibility of our user certificates with a generation and verification script written in Python and a verification app written for Android.

To generate an example SecureABC QR code, run [Generate_QR_Credential.py](Generate_QR_Credential.py), and to verify run either [Verify_QR_Credential.py](Verify_QR_Credential.py) or the Android application which is located in the [SecureABCVerifier](SecureABCVerifier/) directory.

. Credentials, which can be displayed as standard  QR codes, are signed using ECDSA over brainpoolP512r1 and comprise a user photo, name, a unique Certificate ID (CID) number and a validity period. Credentials are completely self-contained and can be verified offline, without interacting with the signer. Revocation is based on distributing revoked CIDs to verifiers.

Example credential verification using our demonstration application:

<p align="center">
	<img src="user_qr.png" alt="Demo SecureABC QR Cerification Application", width=70%>
</p>

## Running the code

The requirements.txt file contains all dependencies.

To run the implmentation you can use the following commands

```
virtualenv venv
source venv/bin/activate
pip install -r requirements.txt
python3 Generate_QR_Credential.py 
python3 Verify_QR_Credential.py
```




changelog.md
# SecureABC Paper Change Log

Here we detail changes to the [SecureABC paper](SecureABC.pdf) which is mirrored here from [arXiv](https://arxiv.org/abs/2005.11833).

* 29 August 2020
 * We have improved our implementation of the system and correspondingly benchmarked its performance in the “implementation and performance” section of our paper. Our implementation now includes an Android application that is able to scan a SecureABC QR code, verify its cryptographic signature and then display the user attributes. Our implementation reveals some of the limitations of QR codes and helps to roadmap the key areas for optimisation of our approach in the future.
 * Following constructive debate about the exclusionary nature of antibody certificates we have rewritten our third general principle to now advocate for proportional and responsible use rather than "no restrictions...".
 * Our security evaluation has been tightened to focus only on the aspects that are within the technical scope of the work. 
 * We have made a number of additional minor improvements throughout the paper.

* 05 June 2020
  * Added new differentially private antibody certificate use case to appendices (in Section 9). This is an extension of the ideas in use case 2 which were developed in collaboration with James Bell.
  * Incorporated early feedback including greater specificity for use case 3, mitigating possible economic discrimination on the basis of requiring a user photograph, clarifying the use of GAN as a source of image only and the purpose of encryption in the app-based authentication protocol.
  * Improved literature review with updated references and greater clarity.
  * Fixes to spelling and grammar.


* 25 May 2020
  * Initial working paper online.


SecureABCVerifier/app/src/test/java/com/example/secureabcverifier/ExampleUnitTest.kt
package com.example.secureabcverifier

import org.junit.Test

import org.junit.Assert.*

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * See [testing documentation](http://d.android.com/tools/testing).
 */
class ExampleUnitTest {
    @Test
    fun addition_isCorrect() {
        assertEquals(4, 2 + 2)
    }
}

Verify_QR_Credential.py
#!/usr/bin/env python3
# Tool to demonstrate the feasibility of encoding a SecureABC
# "AntiBody Certificate" into a standard QR-code.
# C Hicks 16-05-2020
#
# Dependencies:
#   zbar            http://zbar.sourceforge.net
#   opencv-python3  https://pypi.org/project/opencv-python/
#   matplotlib      https://pypi.org/project/matplotlib/
#   OpenSSL         https://pypi.org/project/pyOpenSSL/ (deprecated for crypto)

from pyzbar.pyzbar import decode
import matplotlib.pyplot as plt
import OpenSSL
from OpenSSL import crypto
import cv2
import base64
import io
import time


# QR data class is used to load the QR data payload, extract the certificate
# details and verify the cryptographic signature.
class QR_data():
    def __init__(self, b64_data):
        readPos = 0
        self.raw_data = base64.b64decode(b64_data)
        user_img_length = int.from_bytes(self.raw_data[:2], 'big')
        readPos += 2
        self.user_img_bytes = self.raw_data[readPos:readPos+user_img_length]
        readPos += user_img_length
        user_name_length = int.from_bytes(self.raw_data[readPos:readPos+2], 'big')
        readPos += 2
        self.user_name_bytes = self.raw_data[readPos:readPos+user_name_length]
        readPos += user_name_length
        user_date_length = int.from_bytes(self.raw_data[readPos:readPos+2], 'big')
        readPos += 2
        self.date_bytes = self.raw_data[readPos:readPos+user_date_length]
        readPos += user_date_length
        user_CID_length = int.from_bytes(self.raw_data[readPos:readPos+2], 'big')
        readPos += 2
        self.user_CID_bytes = self.raw_data[readPos:readPos+user_CID_length]
        readPos += user_CID_length
        self.raw_verify_bytes = self.raw_data[0:readPos]
        cert_sign_length = int.from_bytes(self.raw_data[readPos:readPos+2], 'big')
        readPos += 2
        self.cert_sign_bytes = self.raw_data[readPos:readPos+cert_sign_length]

    # Return the user photograph bytes
    def get_user_photo(self):
        return self.user_img_bytes

    # Return the certificate asignee name as a string
    def get_user_name(self):
        return self.user_name_bytes.decode('utf-8')

    # Return the certificate valid date as a string
    def get_user_date(self):
        return self.date_bytes.decode('utf-8')

    # Return the CID number as a string
    def get_user_CID(self):
        return self.user_CID_bytes.decode('utf-8')

    # Return the cryptographic signature on the unsigned certificate bytes
    def get_signature(self):
        return self.cert_sign_bytes

    # Verify the signature on the certificate
    def verify_signature(self, cert):
        tic = time.perf_counter()
        cert_verify = OpenSSL.crypto.verify(cert, self.cert_sign_bytes,
                                                self.raw_verify_bytes, 'sha256')
        toc = time.perf_counter()
        print('Verifying QR signature took {} ms'.format((toc-tic)*1000))
        if cert_verify is None:
            return True
        else:
            return False

#
def main():
    signer_cert_filename = 'OpenSSLKeys/sign_cert.pem'
    qr_code_filename = 'user_qr.png'

    # Load signer certificate public key for verifying user certificate data
    verify_cert_file = open(signer_cert_filename, "r")
    verify_cert = verify_cert_file.read()
    verify_cert_file.close()
    if verify_cert.startswith('-----BEGIN '):
        pubkey = crypto.load_certificate(crypto.FILETYPE_PEM, verify_cert)
    else:
        pubkey = crypto.load_pkcs12(verify_cert).get_certificate()

    # Read QR code
    print('Scanning {} for QR code credential'.format(qr_code_filename))
    qr_data_base64 = decode(cv2.imread(qr_code_filename))[0].data

    # Read certificate data from QR code
    qr = QR_data(qr_data_base64)
    print('QR certificate details: ')
    print('\tName: {}'.format(qr.get_user_name()))
    print('\tDate: {}'.format(qr.get_user_date()))
    print('\t CID: {}'.format(qr.get_user_CID()))
    print('\tSign: {}'.format(base64.encodebytes(qr.get_signature())))

    # Verify signature on certificate data
    print('\t Valid signature: {}'.format(qr.verify_signature(pubkey)))

    # Read and display photo from QR code
    user_photo_stream = io.BytesIO(qr.get_user_photo())
    user_photo = plt.imread(user_photo_stream, 'jpeg')
    plt.imshow(user_photo, cmap='gray')
    plt.axis('off')
    plt.show()

if __name__ == '__main__':
    main()


SecureABCVerifier/app/src/main/java/com/example/secureabcverifier/MainActivity.kt
package com.example.secureabcverifier

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.ImageFormat.*
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.util.Base64.decode
import android.util.Log
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.zxing.*
import com.google.zxing.common.HybridBinarizer
import kotlinx.android.synthetic.main.activity_main.*
import java.io.File
import java.nio.BufferUnderflowException
import java.nio.ByteBuffer
import java.nio.charset.Charset
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

private fun ByteBuffer.toByteArray(): ByteArray {
    rewind()
    val data = ByteArray(remaining())
    get(data)
    return data
}

class QRAnalyzer(private val onQrCodesDetected: (qrCode: Result) -> Unit) : ImageAnalysis.Analyzer {

    private val yuvFormats = mutableListOf(YUV_420_888)

    init {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            yuvFormats.addAll(listOf(YUV_422_888, YUV_444_888))
        }
    }

    private val reader = MultiFormatReader().apply {
        val map = mapOf(
                DecodeHintType.POSSIBLE_FORMATS to arrayListOf(BarcodeFormat.QR_CODE),
                DecodeHintType.TRY_HARDER to true, // Spend more time to try to find a barcode; optimise for accuracy, not speed.
                DecodeHintType.ALLOWED_LENGTHS to intArrayOf(25)
        )
        setHints(map)
    }

    override fun analyze(image: ImageProxy) {
        // We are using YUV format because, ImageProxy internally uses ImageReader to get the image
        // by default ImageReader uses YUV format unless changed.
        if (image.format !in yuvFormats) {
            Log.e("QRCodeAnalyzer", "Expected YUV, now = ${image.format}")
            return
        }

        val data = image.planes[0].buffer.toByteArray()

        val source = PlanarYUVLuminanceSource(
                data,
                image.width,
                image.height,
                0,
                0,
                image.width,
                image.height,
                false
        )

        val binaryBitmap = BinaryBitmap(HybridBinarizer(source))
        try {
            // Whenever reader fails to detect a QR code in image
            // it throws NotFoundException
            val result = reader.decode(binaryBitmap)
            onQrCodesDetected(result)
        } catch (e: NotFoundException) {
            // No need to do anything unless we DO find a QR code
            //e.printStackTrace()
        }
        image.close()
    }
}

const val SECUREABC_QR_DATA = "com.example.secureabcverifier.QRDATA"
const val SECUREABC_VERIFY_DATA = "com.example.secureabcverifier.VERIFYDATA"
const val SECUREABC_VERIFY_SIGNATURE = "com.example.secureabcverifier.VERIFYSIGNATURE"
const val SECUREABC_QR_PHOTO = "com.example.secureabcverifier.QRPHOTO"

class MainActivity : AppCompatActivity() {
    private var imageCapture: ImageCapture? = null
    private var isProcessing = false

    private lateinit var outputDirectory: File
    private lateinit var cameraExecutor: ExecutorService

    override fun onRequestPermissionsResult(
            requestCode: Int, permissions: Array<String>, grantResults:
            IntArray) {
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera()
            } else {
                Toast.makeText(this,
                        "Permissions not granted by the user.",
                        Toast.LENGTH_SHORT).show()
                finish()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Request camera permissions
        if (allPermissionsGranted()) {
            startCamera()
        } else {
            ActivityCompat.requestPermissions(
                    this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS)
        }

        outputDirectory = getOutputDirectory()

        cameraExecutor = Executors.newSingleThreadExecutor()
    }


    private fun startCamera() {
        val cameraProviderFuture = ProcessCameraProvider.getInstance(this)

        cameraProviderFuture.addListener(Runnable {
            // Used to bind the lifecycle of cameras to the lifecycle owner
            val cameraProvider: ProcessCameraProvider = cameraProviderFuture.get()

            // Preview
            val preview = Preview.Builder()
                    .build()
                    .also {
                        it.setSurfaceProvider(viewFinder.createSurfaceProvider())
                    }

            val imageAnalyzer = ImageAnalysis.Builder()
                    //.setTargetResolution(Size(previewView.width, previewView.height))
                    .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                    .build()
                    .also {
                        it.setAnalyzer(cameraExecutor, QRAnalyzer { qrResult ->
                            if(!isProcessing) {
                                isProcessing = true

                                Log.d("QRCodeAnalyzer", "Barcode scanned (${qrResult.numBits} bits): ${qrResult.text}")

                                val messagelength = qrResult.numBits
                                if (messagelength >= 18672) {
                                    try {

                                        val qrData = decode(qrResult.text, android.util.Base64.DEFAULT)
                                        Log.d("QRCodeAnalyzer", "(${qrData.size} bytes QR payload..)")

                                        var readOffset = 0
                                        val imglen = ByteBuffer.wrap(qrData, readOffset, 2).getShort().toInt()
                                        readOffset += 2
                                        Log.d("QRCodeAnalyzer", "(imgLen = ${imglen} )")
                                        val imgbytes = qrData.copyOfRange(readOffset, readOffset+imglen)
                                        readOffset += imglen
                                        Log.d("QRCodeAnalyzer", "(readOffset = ${readOffset} )")
                                        val namelen = ByteBuffer.wrap(qrData, readOffset, 2).getShort().toInt()
                                        readOffset += 2
                                        Log.d("QRCodeAnalyzer", "(namelen = ${namelen} )")
                                        val username = String(qrData.copyOfRange(readOffset, readOffset+namelen), Charsets.UTF_8)
                                        readOffset+=namelen
                                        Log.d("QRCodeAnalyzer", "(name = ${username})")
                                        val datelen = ByteBuffer.wrap(qrData, readOffset, 2).getShort().toInt()
                                        readOffset += 2
                                        Log.d("QRCodeAnalyzer", "(datelen = ${datelen} )")
                                        val userdate = String(qrData.copyOfRange(readOffset, readOffset+datelen), Charsets.UTF_8)
                                        readOffset += datelen
                                        Log.d("QRCodeAnalyzer", "(date = ${userdate} )")
                                        val CIDlen = ByteBuffer.wrap(qrData, readOffset, 2).getShort().toInt()
                                        readOffset += 2
                                        Log.d("QRCodeAnalyzer", "(CIDlen = ${CIDlen} )")
                                        val userCID = String(qrData.copyOfRange(readOffset, readOffset+CIDlen), Charsets.UTF_8)
                                        readOffset += CIDlen
                                        Log.d("QRCodeAnalyzer", "(date = ${userCID} )")

                                        // Read signature
                                        val rawVerifyBytes = qrData.copyOfRange(0, readOffset)
                                        val certSigLen = ByteBuffer.wrap(qrData, readOffset, 2).getShort().toInt()
                                        readOffset += 2
                                        Log.d("QRCodeAnalyzer", "(cert sig length = ${certSigLen} )")
                                        val sigBytes = qrData.copyOfRange(readOffset, readOffset+certSigLen)


                                        val qrdata = arrayOf(username, userdate, userCID)
                                        val intent = Intent(this, VerifyActivity::class.java).apply {
                                            putExtra(SECUREABC_QR_PHOTO, imgbytes)
                                            putExtra(SECUREABC_QR_DATA, qrdata)
                                            putExtra(SECUREABC_VERIFY_DATA, rawVerifyBytes)
                                            putExtra(SECUREABC_VERIFY_SIGNATURE, sigBytes)
                                        }

                                        startActivity(intent)
                                    } catch(e: IllegalArgumentException) {
                                        // Bad base64 in input; try again
                                        isProcessing=false
                                    } catch(e: BufferUnderflowException) {
                                        // Read bad lengths from QR code and try to read past buffer
                                        isProcessing=false
                                    } catch(e: IndexOutOfBoundsException) {
                                        // Read too far
                                        isProcessing=false
                                    }
                                } else{
                                    isProcessing=false
                                }


                            }
                        })
                    }


            imageCapture = ImageCapture.Builder()
                    .build()

            // Select back camera as a default
            val cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA

            try {
                // Unbind use cases before rebinding
                cameraProvider.unbindAll()

                // Bind use cases to camera
                cameraProvider.bindToLifecycle(
                        this, cameraSelector, preview, imageCapture, imageAnalyzer)

            } catch (exc: Exception) {
                Log.e(TAG, "Use case binding failed", exc)
            }

        }, ContextCompat.getMainExecutor(this))
    }

    private fun takePhoto() {
        // Get a stable reference of the modifiable image capture use case
        val imageCapture = imageCapture ?: return

        // Create time-stamped output file to hold the image
        val photoFile = File(
                outputDirectory,
                SimpleDateFormat(FILENAME_FORMAT, Locale.US
                ).format(System.currentTimeMillis()) + ".jpg")

        // Create output options object which contains file + metadata
        val outputOptions = ImageCapture.OutputFileOptions.Builder(photoFile).build()

        // Set up image capture listener, which is triggered after photo has
        // been taken
        imageCapture.takePicture(
                outputOptions, ContextCompat.getMainExecutor(this), object : ImageCapture.OnImageSavedCallback {
            override fun onError(exc: ImageCaptureException) {
                Log.e(TAG, "Photo capture failed: ${exc.message}", exc)
            }

            override fun onImageSaved(output: ImageCapture.OutputFileResults) {
                val savedUri = Uri.fromFile(photoFile)
                val msg = "Photo capture succeeded: $savedUri"
                Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
                Log.d(TAG, msg)
            }
        })
    }

    private fun allPermissionsGranted() = REQUIRED_PERMISSIONS.all {
        ContextCompat.checkSelfPermission(
                baseContext, it) == PackageManager.PERMISSION_GRANTED
    }

    private fun getOutputDirectory(): File {
        val mediaDir = externalMediaDirs.firstOrNull()?.let {
            File(it, resources.getString(R.string.app_name)).apply { mkdirs() } }
        return if (mediaDir != null && mediaDir.exists())
            mediaDir else filesDir
    }

    override fun onResume() {
        super.onResume()
        isProcessing=false
    }

    override fun onDestroy() {
        super.onDestroy()
        cameraExecutor.shutdown()
    }

    companion object {
        private const val TAG = "CameraXBasic"
        private const val FILENAME_FORMAT = "yyyy-MM-dd-HH-mm-ss-SSS"
        private const val REQUEST_CODE_PERMISSIONS = 10
        private val REQUIRED_PERMISSIONS = arrayOf(Manifest.permission.CAMERA)
    }
}


SecureABCVerifier/app/src/main/java/com/example/secureabcverifier/VerifyActivity.kt
package com.example.secureabcverifier

import android.graphics.BitmapFactory
import android.graphics.Color
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.ImageView
import android.widget.TextView
import java.io.ByteArrayInputStream
import java.io.FileInputStream
import java.security.InvalidKeyException
import java.security.KeyStore
import java.security.NoSuchAlgorithmException

import java.security.Signature
import java.security.cert.Certificate
import java.security.cert.CertificateFactory
import java.util.*
import javax.net.ssl.TrustManagerFactory

class VerifyActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_verify)

        var certStatus = false

        // Get the Intent that started this activity and extract the string (b64 data)
        val qrdata  = intent.getStringArrayExtra(SECUREABC_QR_DATA)
        val verifydata = intent.getByteArrayExtra(SECUREABC_VERIFY_DATA)
        val verifySig = intent.getByteArrayExtra(SECUREABC_VERIFY_SIGNATURE)
        val qrphoto = intent.getByteArrayExtra(SECUREABC_QR_PHOTO)

        // Load certificate for verification
        val certFactory = CertificateFactory.getInstance("X.509")
        val certStream =  resources.openRawResource(R.raw.sign_cert)
        val cert = certFactory.generateCertificate(certStream)
        Log.d("QRCodeAnalyzer", "certificate: ${cert.publicKey}")

        // Verify signature
        try {
            val signature = Signature.getInstance("SHA256withECDSA")
            signature.initVerify(cert)
            val benchmark = Date()
            signature.update(verifydata)
            certStatus = signature.verify(verifySig)
            val duration = (Date().time - benchmark.time)
            Log.d("QRCodeAnalyzer", "Time to verify: ${duration} ms")

        } catch (e: InvalidKeyException) {
            Log.d("QRCodeAnalyzer", "Error loading certificate: ${e}")
        } catch (e: NoSuchAlgorithmException) {
            Log.d("QRCodeAnalyzer", "Error verifying signature: ${e}")
        }

        // Set photo
        var photoStream = ByteArrayInputStream(qrphoto)
        val imageView = findViewById<ImageView>(R.id.imageView).apply {
            setImageBitmap(BitmapFactory.decodeStream(photoStream))
        }

        // Capture the layout's TextView's and set the strings using QR data payload
        val nameView = findViewById<TextView>(R.id.nameView).apply {
            text = "Name: " + qrdata[0] //Integer.toString(user_name_length)
        }
        val dateView = findViewById<TextView>(R.id.dateView).apply {
            text = "Validity: " + qrdata[1] //Integer.toString(user_name_length)
        }
        val cidView = findViewById<TextView>(R.id.cidView).apply {
            text = "CID: " + qrdata[2] //Integer.toString(user_name_length)
        }
        val validView = findViewById<TextView>(R.id.validView).apply {
            if (certStatus) {
                setTextColor(Color.GREEN)
                text = "Certificate: Valid"
            } else {
                setTextColor(Color.RED)
                text = "Certificate: NOT VALID"
            }

        }

    }
}

Generate_QR_Credential.py
#!/usr/bin/env python3
# Tool to demonstrate the feasibility of encoding a SecureABC
# "AntiBody Certificate" into a standard QR-code.
# C Hicks 16-05-2020
#
# Dependencies:
#   qrcode 6.1  https://pypi.org/project/qrcode/
#   matplotlib  https://pypi.org/project/matplotlib/
#   OpenSSL     https://pypi.org/project/pyOpenSSL/ (deprecated for crypto)

import qrcode
import matplotlib.pyplot as plt
import base64
import OpenSSL
from OpenSSL import crypto
import time

# Certificate data object
class CertificateData:

    # Initialise certificate data
    def __init__(self, user_img, user_name, user_date, user_CID):
        self.user_img = user_img  # user_img_arr = np.asarray(user_img)
        self.user_name = user_name
        self.user_date = user_date
        self.user_CID = user_CID
        self.cert_bytes = self.getByteArray()

    # Return the certificate data as a byte array with variable field lengths
    def getByteArray(self):
        user_img_length = len(self.user_img).to_bytes(2, 'big')
        user_img_bytes = self.user_img
        user_name_len = len(self.user_name).to_bytes(2, 'big')
        user_name_bytes = bytearray(self.user_name, 'utf-8')
        user_date_len = len(self.user_date).to_bytes(2, 'big')
        user_date_bytes = bytearray(self.user_date, 'utf-8')
        user_CID_len = len(self.user_CID).to_bytes(2, 'big')
        user_CID_bytes = bytearray(self.user_CID, 'utf-8')

        #
        cert_bytes = user_img_length + user_img_bytes +\
                        user_name_len + user_name_bytes +\
                            user_date_len + user_date_bytes +\
                                user_CID_len + user_CID_bytes

        return cert_bytes

    # Signs the certificate data using pkey and then returns the certificate
    # data, plus the signature, as a bytearray
    def getSignedCertificateByteArray(self, pkey):
        tic = time.perf_counter()
        cert_sign = OpenSSL.crypto.sign(pkey, self.cert_bytes, "sha256")
        toc = time.perf_counter()
        print('{} seconds to generate signature'.format((toc-tic)*1000))
        cert_sign_len = len(cert_sign).to_bytes(2, 'big')
        cert_sign_bytes = bytearray(cert_sign)

        return self.cert_bytes + cert_sign_len + cert_sign_bytes

def main():
    signer_pkey_filename = 'OpenSSLKeys/sign_key.pem'
    user_image_filename = 'user_bw.jpeg'
    qr_code_filename = 'user_qr.png'

    # Define QR code specification
    qr = qrcode.QRCode(
        version=40,
        error_correction=qrcode.constants.ERROR_CORRECT_L,
        box_size=100,
        border=4,
    )

    # Load private key for signing user data
    key_file = open(signer_pkey_filename, "r")
    key = key_file.read()
    key_file.close()
    if key.startswith('-----BEGIN '):
        pkey = crypto.load_privatekey(crypto.FILETYPE_PEM, key)
    else:
        pkey = crypto.load_pkcs12(key).get_privatekey()

    tic = time.perf_counter()

    # Create user data
    user_img = open(user_image_filename, "rb").read()
    user_name = "Alice Doe"
    user_date = "16/08/2020 - 16/11/2020"
    user_CID = "0x1fc60e1a4e238ac6cce9d79097a268af"
    user_certData = CertificateData(user_img, user_name, user_date, user_CID)

    # Get signed user certificate and build QR code
    signed_user_cert = user_certData.getSignedCertificateByteArray(pkey)

    # qr.add_data(base64.b64encode(signed_user_cert))
    qr.add_data(base64.b64encode(signed_user_cert))

    try:
        qr.make(fit=False)
    except qrcode.exceptions.DataOverflowError:
        print('User data too big for QR code. Please optimise image.')
        exit()

    qr_img = qr.make_image(fill_color="black", back_color="white")
    plt.imshow(qr_img, cmap='gray')
    plt.axis('off')
    plt.savefig(qr_code_filename, bbox_inches='tight')
    toc = time.perf_counter()

    print('{} seconds to generate complete QR code output'.format((toc-tic)*1000))
    print('User QR code output to {}'.format(qr_code_filename))
    print('Done.')


if __name__ == '__main__':
    main()
